implement support for inherited properties

Ashymonth opened the pull request:
Add the ability to generate inherited properties
In order not to break compatibility with previous versions and to allow the user to choose for himself how to act with the properties that are in the base class, the ```InheritedColumnsOrdering ``` attribute was added.  You can use it to set 3 types of behavior for inherited properties. 
 - ```IgnoreInheritedProperties``` - properties from base class will be igonred. This is default behaviour.
 - ```StartFromInheritedProperties ```- Property generation will start with the properties of the base class.
 - ```StartFromClassProperties ```- Property generation will start from the current class

Ashymonth commented:
i also added test for new staff. I tried to follow your code style. If you want to change smth i can do it. 
I also added using ```classType.GetMembers().OfType<IPropertySymbol>();```, but if you say to revert  I'll do it

sveinungf commented:
Thanks for the PR.
1. It looks like the code handles properties from the direct base class. But what if there are two or more levels of inheritance?
2. I'm a bit unsure about the attribute name `InheritedColumnOrdering`. I think in some cases you would want to include the properties from the base class, but not necessarily set the column ordering. For example if all properties have the `ColumnOrderAttribute` set, then having to also decide the ordering for this attribute can be a bit confusing. Maybe the name could basically be something that means "take the base class into account". The column ordering could be an optional parameter for the attribute.
3. Ideally the name should make it clear that it sets the default ordering. Any use of the `ColumnOrderAttribute` can be used to override the default order.

Perhaps a way to solve these issues could be to have the attribute named `InheritColumns`, with optional parameter `DefaultOrder` which could be either `InheritedColumnOrder.InheritedColumnsFirst` or `InheritedColumnOrder.InheritedColumnsLast`. If there are more levels of inheritance, you would also have to set the attribute on the base class, and optionally the parameter as well.

An example of how this could work:

```csharp
public class Animal
{
    public DateTime DateOfBirth { get; set; }
}

[InheritColumns(DefaultOrder = InheritedColumnOrder.InheritedColumnsLast)]
public class Mammal : Animal
{
    public bool CanWalk { get; set; }
}

[InheritColumns] // Defaults to InheritedColumnsFirst
public class Dog : Mammal
{
    public string Breed { get; set; }
}
```

The order of the columns here should be: CanWalk, DateOfBirth, Breed.

Ashymonth commented:
I fixed the deep inheritance issues and updated the attribute name. But I'm not sure about your example.
If you do so, then the class is `Animal` it will not have the `InheritColumns` attribute, but it will still inherit properties from base classes. I think if you want to inherit properties from base classes, you need to set this attribute manually, otherwise you won't be able to disable this behavior.
What do you think?

sveinungf commented:
In the example `Animal` doesn't have the attribute because it doesn't have a base class, only derived classes. So it doesn't have anything to inherit.

Otherwise I agree, the attribute must be there to inherit properties. If the attribute was removed from the `Mammal` class in the example, then the property from the `Animal` class would be ignored and there would only be 2 columns: CanWalk and Breed. If the attribute was removed from the `Dog` class, then the behavior is the same as today with Breed being the only column.

Ashymonth commented:
Yeah you're right.
 The order of the columns here should be: CanWalk, DateOfBirth, Breed confused me a little bit, sorry
Then it my last commits it implemented correctly

Ashymonth commented:
Hi. Any updates ?

Ashymonth commented on SpreadCheetah/SourceGeneration/InheritColumnsAttribute.cs:
```diff
@@ -0,0 +1,17 @@
+    /// Set default order for inherited properties.
+    /// </summary>
+    /// <param name="defaultOrder"></param>
+    public InheritColumnsAttribute(InheritedColumnOrder defaultOrder = InheritedColumnOrder.InheritedColumnsFirst)
```
sure

Ashymonth commented on SpreadCheetah/SourceGeneration/InheritColumnsAttribute.cs:
```diff
@@ -0,0 +1,17 @@
+/// <summary>
+/// Instructs the SpreadCheetah source generator how to handle inherited properties.
+/// </summary>
+[AttributeUsage(AttributeTargets.Class)]
```
added

sveinungf commented:
This looks very good, thank you!
